//------------------------------------------------------------------------------
// {C} Copyright 2020 Pensando Systems Inc. All rights reserved
//
// protobuf specification for Nicmgr Upgrade Backup/Restore object
//------------------------------------------------------------------------------

syntax = "proto3";
package pds.nicmgr;

import "meta/meta.proto";
import "types.proto";

message EthDeviceRdmaSpec {
  bool   Enable   = 1;
  uint32 PteCount = 2;
  uint32 KeyCount = 3;
  uint32 AHCount  = 4;
  uint32 SQCount  = 5;
  uint32 RQCount  = 6;
  uint32 CQCount  = 7;
  uint32 EQCount  = 8;
  uint32 AQCount  = 9;
  uint32 PidCount = 10;
}

message EthDevicePcieSpec {
  string OpromType = 1;
  uint32 Port      = 2;
  uint32 TotalVfs  = 3;
}

message EthDeviceSpec {
  // device
  string            Name             = 1;
  uint64            DevUuid          = 2;
  string            EthType          = 3;
  bool              HostDev          = 4;
  uint32            BarMapSize       = 5;
  // network
  uint32            UplinkPortNumber = 6;
  string            QosGroup         = 7;
  // resource counts
  uint32            LifCount         = 8;
  uint32            RxQCount         = 9;
  uint32            TxQCount         = 10;
  uint32            EQCount          = 11;
  uint32            AdminQCount      = 12;
  uint32            IntrCount        = 13;
  uint64            MacAddr          = 14;
  // pcie spec
  EthDevicePcieSpec PcieSpec         = 15;
  // rdma spec
  EthDeviceRdmaSpec RdmaSpec         = 16;
}

message EthDeviceStatus {
  uint32 LifBase      = 1;
  uint32 IntrBase     = 2;
  uint64 RegsMemAddr  = 3;
  uint64 PortInfoAddr = 4;
  uint64 CmbMemAddr   = 5;
  uint64 CmbMemSize   = 6;
  uint64 RomMemAddr   = 7;
  uint64 RomMemSize   = 8;
}

message EthDeviceStats {

}

message EthDevice {
  meta.TypeMeta   TypeMeta = 1 [(gogoproto.embed) = true, (gogoproto.jsontag) = ",inline"];
  meta.ObjMeta    ObjMeta  = 2 [(gogoproto.embed) = true, (gogoproto.jsontag) = "meta,omitempty"];
  EthDeviceSpec   Spec     = 3;
  EthDeviceStatus Status   = 4;
  EthDeviceStats  Stats    = 5;
}
